### ISSUE TITLE
Assigning to attribute introduces unknown in the right-hand-side expression

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#3027]
With the latest Pyright version there's a weird bug where sometimes the control-flow analysis seems to get confused and produce unknowns when assigning an expression to an attribute. I can reproduce the problem by using the walrus operator in a generator expression. There's no error when you assign to a variable or an index, it's just for attributes.

```py
@dataclass
class Foo:
    input: str
    output: str

def minify1(foo: Foo):
    # Argument type is partially unknown
    #   Argument corresponds to parameter "__iterable" in function "join"
    #   Argument type is "Generator[Unknown, None, None]"
    foo.output = "\n".join(
        stripped for line in foo.input.splitlines() if (stripped := line.strip())
    )

def minify2(foo: Foo):
    # This works fine, no errors
    output = "\n".join(
        stripped for line in foo.input.splitlines() if (stripped := line.strip())
    )
    foo.output = output
```

**Comment by erictraut** [comment]
Is the error deterministic? Or does it seem to come and go depending on the edits you make in the file? Pyright can evaluate expressions in any order (e.g. to service a hover text request immediately after an edit). It is supposed to always produce the same results regardless of evaluation order, but sometimes bugs in the evaluator produce different (seemingly nondeterministic) results. I'm wondering if that's the behavior you're seeing here.

I'm not able to repro any error using the code in your example above.

Can you provide more details about your configuration? Are you using the command-line version of pyright? the language server version? The VS Code extension? Which version of Python are you using? Which platform?

**Comment by vberlier** [comment]
Yeah it seems to be deterministic, I can paste the example anywhere and I'll get the error. I'm using the command-line with Pyright `1.1.220` with strict mode. It also shows up in vscode because I think Pylance uses the version in node_modules. I'm using WSL with python 3.8.

The file causing problem is here https://github.com/mcbeet/beet/blob/main/beet/contrib/minify_function.py

**Comment by erictraut** [comment]
Ah, I forgot to enable strict mode. I'm able to repro it now. I'll look into it.

**Comment by erictraut** [comment]
Thanks for reporting this issue. It will be fixed in the next release.

**Comment by vberlier** [comment]
Thanks for the quick fix!

**Comment by erictraut** [comment]
This is included in pyright 1.1.221, which I just published. It will also be included in the next release of pylance.
